jormungandr/src/secure/enclave.rs

```diff
@@ -152,3 +231,118 @@ impl Enclave {
         }
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use chain_crypto::{Blake2b256, Ed25519, SecretKey};
+    use chain_impl_mockchain::certificate::PoolId;
+    use chain_impl_mockchain::fragment::Fragment::PoolRegistration;
+    use chain_impl_mockchain::leadership::{BftLeader, GenesisLeader};
+    use rand_core;
+    use tokio02 as tokio;
+
+    #[tokio::test]
+    async fn enclave_add_different_bft_leaders() {
+        let mut enclave = Enclave::new();
+        let rng = rand_core::OsRng;
+        let leader1 = Leader {
+            bft_leader: Some(BftLeader {
+                sig_key: SecretKey::generate(rng),
+            }),
+            genesis_leader: None,
+        };
+        let leader2 = Leader {
+            bft_leader: Some(BftLeader {
+                sig_key: SecretKey::generate(rng),
+            }),
+            genesis_leader: None,
+        };
+        let init_leader_id = LeaderId::new();
+        let fst_id = init_leader_id.next();
+        let snd_id = fst_id.next();
+        assert_eq!(enclave.add_leader(leader1).await, fst_id);
+        assert_eq!(enclave.add_leader(leader2).await, snd_id);
+        let leaders_data = &enclave.leaders_data.read().await;
+        assert_eq!(leaders_data.leaders.len(), 2);
+        assert_eq!(leaders_data.added_leaders_cache.len(), 2);
+    }
+
+    #[tokio::test]
+    async fn enclave_add_duplicated_bft_leaders() {
+        let mut enclave = Enclave::new();
+        let secret_key = SecretKey::generate(rand_core::OsRng);
+        let leader1 = Leader {
+            bft_leader: Some(BftLeader {
+                sig_key: secret_key.clone(),
+            }),
+            genesis_leader: None,
+        };
+        let leader2 = Leader {
+            bft_leader: Some(BftLeader {
+                sig_key: secret_key.clone(),
+            }),
+            genesis_leader: None,
+        };
+        // Both leaders are different instances of the same data, adding both of them should return the same id
+        assert_eq!(
+            enclave.add_leader(leader1).await,
+            enclave.add_leader(leader2).await
+        );
+        // Just one it is really added
+        let leaders_data = &enclave.leaders_data.read().await;
+        assert_eq!(leaders_data.leaders.len(), 1);
+        assert_eq!(leaders_data.added_leaders_cache.len(), 1);
+    }
+
+    //    #[tokio::test]
+    //    async fn enclave_add_different_genesis_leaders() {
+    //        let mut enclave = Enclave::new();
+    //        let rng = rand_core::OsRng;
+    //        let leader1 = Leader {
+    //            bft_leader: None,
+    //            genesis_leader: Some( GenesisLeader {
```

Comment:
I'm having some trouble creating a `GenesisLeader`. I couldn't find any reference or example on how to do it for testing. Any help on how to do it is more than welcome. @NicolasDP, any tip?